Automattic/wp-calypso
# Google Analytics 4 For WPcom

#### Proposed Changes

* Implementing Google Analytics 4 for WPcom
* Will handle page views, generic events, and purchase tracking.
* Related to: pau2Xa-48B-p2 
* Basically, the code is made more generic from the previous implementation for Jetpack Cloud. This allows us to re-use the functionality for WPcom (and potentially other sites down the line).

#### Notes
* Purchases for Jetpack and WPcom are fired on the WPcom implementation. This is in line with the current implementation of Enhanced Ecommerce tracking in Universal Analytics. I've updated the PR to reflect this, and making the product brand independent of the GA4 tracking environment. This also takes into account some "edge cases" e.g. if the cart has both a WPcom and Jetpack product, both will be parsed. One example would be https://wordpress.com/checkout/jetpack_search,premium/site_slug. Which will generate the cart items below:

```javascript
(2) [{…}, {…}]
0: {product_id: 1003, billing_plan_id: '16615', product_name: 'WordPress.com Premium', product_name_en: 'WordPress.com Premium', product_slug: 'value_bundle', …}
1: {product_id: 800, billing_plan_id: '20459', product_name: 'Jetpack Search', product_name_en: 'Jetpack Search', product_slug: 'wpcom_search', …}
length: 2
[[Prototype]]: Array(0)
```

#### Testing Instructions
It may be helpful to open the network tab in developer tools, and from there filter on `G-1H4VG5F5JF` (which is the measurement ID we're using for GA4).

* Sandbox store (purchase events won't fire when using free credits, as cost < $0.01).
* Allow ad tracking using config flag `?flags=gdpr-banner,google-analytics,ad-tracking`.
    * If you're in EU, you can accept the cookie banner -- or set the `sensitive_pixel_option` cookie manually. 
* Check that gtag.js is initializing correctly. You should see a call with the parameter `config`.
* Check that page views are firing (event name `page_view`).
* Buy a plan, verify that purchase is triggering (event name `purchase`).
* Automated tests pass.
* No TS/console errors.
* If you have access to the GA4 property, you can also verify that the purchase is showing under real-time. 
* **Verify that the tracking works on Jetpack Cloud, by going through the same test procedure as above.**

#### Pre-merge Checklist

Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.


Resolves Automattic/martech#827 Automattic/martech#831 Automattic/martech#832
